### PR TITLE
[rush] Replace the AWS SDK packages with the S3 REST API.

### PR DESCRIPTION
**Pull request by iclanton**
## Summary

This PR fixes https://github.com/microsoft/rushstack/issues/2547 by removing the AWS npm packages as dependencies of Rush and replaces them with use of the S3 REST API.

## How it was tested

Tested by building the rushstack repo with a S3 bucket configured as the cloud cache provider.

### CODE REVIEW

**Comment by iclanton** [comment]
@raihle - you're more experienced with S3 and the Rush S3 code than I am. Would you mind taking a look through this PR?

File: apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  private async _makeRequestAsync(
+    verb: 'GET' | 'PUT',
+    objectName: string,
+    body?: Buffer
+  ): Promise<fetch.Response> {
+    const isoDateString: IIsoDateString = this._getIsoDateString();
+    const bodyHash: string = this._getSha256(body);
+
+    // Compute the authorization header. See https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+    const host: string = `${this._s3Bucket}.s3.amazonaws.com`;
```
Comment (octogonz) [inline apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts:93]: What ensures that `_s3Bucket` does not contain any invalid characters?

File: apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    const signedHeaderNames: string = `${HOST_HEADER_NAME};${CONTENT_HASH_HEADER_NAME};${DATE_HEADER_NAME}`;
+    const canonicalRequest: string = [
```
Comment (octogonz) [inline apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts:95]: Could you add a code comment giving an example value for the `canonicalRequest` string that is assembled by this pattern?

File: apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+      `${CONTENT_HASH_HEADER_NAME}:${bodyHash}`,
+      `${DATE_HEADER_NAME}:${isoDateString.dateTime}`,
+      '',
+      signedHeaderNames,
+      bodyHash
+    ].join('\n');
+    const canonicalRequestHash: string = this._getSha256(canonicalRequest);
+
+    const scope: string = `${isoDateString.date}/${this._s3Region}/s3/aws4_request`;
+    const stringToSign: string = [
```
Comment (octogonz) [inline apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts:109]: Same here; could you add a code comment showing a typical value for `stringToSign`?

File: apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+  }
+
+  private _getSha256(data?: string | Buffer): string {
+    if (data) {
+      const hash: crypto.Hash = crypto.createHash('sha256');
+      hash.update(data);
+      return hash.digest('hex');
+    } else {
+      return 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855';
```
Comment (octogonz) [inline apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts:163]: What is this magic value?  At least add some docs explaining where it came from

File: apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    }
+  }
+
+  private _getIsoDateString(date: Date = new Date()): IIsoDateString {
+    let dateString: string = date.toISOString();
```
Comment (octogonz) [inline apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts:168]: Show an example value for `dateString` in a code comment, so we know what the `replace()` is supposed to be matching

File: apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    dateString = dateString.replace(/[-:]/g, ''); // Remove separator characters
+    dateString = dateString.substring(0, 15); // Drop milliseconds
```
Comment (octogonz) [inline apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts:170]: Add a code comment showing an example final value for `dateString` after these manipulations

File: apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+    return {
+      dateTime: `${dateString}Z`,
+      date: dateString.substring(0, 8)
```
Comment (octogonz) [inline apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts:174]: Add a code comment showing an example final value for `dateTime`/`date` after these manipulations.

File: apps/rush-lib/package.json [hunk 8]
```diff
@@ -18,7 +18,6 @@
   },
   "license": "MIT",
   "dependencies": {
-    "@aws-sdk/client-s3": "~3.3.0",
```
Comment (octogonz) [inline apps/rush-lib/package.json:21]: Nice!  I'm impressed we were able to eliminate this huge dependency with such a relatively small amount of custom code. 😎

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
@iclanton thank you so much for doing this work!

**Pull request merged**